contractcourt/htlc_timeout_resolver.go

```diff
@@ -67,6 +71,154 @@ func (h *htlcTimeoutResolver) ResolverKey() []byte {
 	return key[:]
 }
 
+const (
+	// expectedRemoteWitnessSuccessSize is the expected size of the witness
+	// on the remote commitment transaction for an outgoing HTLC that is
+	// swept on-chain by them with pre-image.
+	expectedRemoteWitnessSuccessSize = 5
+
+	// remotePreimageIndex index within the witness on the remote
+	// commitment transaction that will hold they pre-image if they go to
+	// sweep it on chain.
+	remotePreimageIndex = 3
+
+	// localPreimageIndex is the index within the witness on the local
+	// commitment transaction for an outgoing HTLC that will hold the
+	// pre-image if the remote party sweeps it.
+	localPreimageIndex = 1
+)
+
+// claimCleanUp is a helper method that's called once the HTLC output is spent
+// by the remote party. It'll extract the preimage, add it to the global cache,
+// and finally send the appropriate clean up message.
+func (h *htlcTimeoutResolver) claimCleanUp(commitSpend *chainntnfs.SpendDetail) (ContractResolver, error) {
+	// Depending on if this is our commitment or not, then we'll be looking
+	// for a different witness pattern.
+	spenderIndex := commitSpend.SpenderInputIndex
+	spendingInput := commitSpend.SpendingTx.TxIn[spenderIndex]
+
+	log.Infof("%T(%v): extracting preimage! remote party spent "+
+		"HTLC with tx=%v", h, h.htlcResolution.ClaimOutpoint,
+		spew.Sdump(commitSpend.SpendingTx))
+
+	// If this is the remote party's commitment, then we'll be looking for
+	// them to spend using the second-level success transaction.
+	var preimageBytes []byte
+	if h.htlcResolution.SignedTimeoutTx == nil {
+		// The witness stack when the remote party sweeps the output to
+		// them looks like:
+		//
+		//  * <0> <sender sig> <recvr sig> <preimage> <witness script>
+		preimageBytes = spendingInput.Witness[remotePreimageIndex]
```

Comment:
Can this give an out of bounds error if for some reason we are running with outdated state and the timeout tx was already published?

Reply:
The `isSuccessSpend` will guard against that since we assert the length of the witness is the expected size. For out of date states, if we knew about the HTLC at the time the state was created, then the timeout transaction will still be populated as that's a matter of the remote vs local commitment.

Reply:
`isSuccessSpend` isn't called in the outgoing contest resolver.